### PR TITLE
Componentizes Surgery-Initiation Devices

### PR DESCRIPTION
**Pull request by ExcessiveUseOfCobblestone**
## About The Pull Request

Converts the act of initiating(/cancelling) surgery to component, allowing for any item to support surgeries. Also supports callbacks to items after selecting a surgery (limited use drapes or something).

## Why It's Good For The Game

Code Improvement

## Changelog

No changes

### CODE REVIEW

**Comment by WarlockD** [comment]
Humm you know, this might work for tools.  I had to freaking redo my fork so might do one for basic tools

**Comment by Fox-McCloud** [comment]
I never understood why it required a bedsheet or drapes---wouldn't it make more sense to tie it with something *sharp*?

**Comment by WarlockD** [comment]
I think it was because the pattern of surgerys were so similar that we needed something to kickstart.  Paradise surgerys don't do this but all their tool patterns are different, but it limits the kinds of surgery's they can do.

**Comment by Jared-Fogle** [comment]
Drapes also make it possible to use an operating computer to guide yourself on how to do the surgery in world.

File: code/datums/components/surgery_initiator.dm [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	after_select_cb = _after_select_cb
+
+/datum/component/surgery_initiator/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ITEM_ATTACK, .proc/initiate_surgery_moment)
+
+/datum/component/surgery_initiator/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_ITEM_ATTACK)
+
+
+	/**
```
Comment (Rohesie) [inline code/datums/components/surgery_initiator.dm:24]:
```suggestion
/datum/component/surgery_initiator/UnregisterFromParent()
	UnregisterSignal(parent, COMSIG_ITEM_ATTACK)


/datum/component/surgery_initiator/Destroy()
	if(after_select_cb)
		QDEL_NULL(after_select_cb)
	return ..()


	/**
```
Some callbacks keep references to things in their `object` var and must be cleaned.

File: code/datums/components/surgery_initiator.dm [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+				user.visible_message("<span class='notice'>[user] drapes [parent] over [M]'s [parse_zone(selected_zone)] to prepare for surgery.</span>", \
+					"<span class='notice'>You drape [parent] over [M]'s [parse_zone(selected_zone)] to prepare for \an [procedure.name].</span>")
+
+				log_combat(user, M, "operated on", null, "(OPERATION TYPE: [procedure.name]) (TARGET AREA: [selected_zone])")
+				after_select_cb?.Invoke()
+			else
+				to_chat(user, "<span class='warning'>You need to expose [M]'s [parse_zone(selected_zone)] first!</span>")
+
+	else if(!current_surgery.step_in_progress)
+		attempt_cancel_surgery(current_surgery, parent, M, user)
```
Comment (Rohesie) [inline code/datums/components/surgery_initiator.dm:110]:
```suggestion
	var/mob/living/living_target = target
	var/mob/living/carbon/carbon_target
	var/obj/item/bodypart/affecting
	var/selected_zone = user.zone_selected
	. = COMPONENT_ITEM_NO_ATTACK

	if(iscarbon(living_target))
		carbon_target = living_target
		affecting = carbon_target.get_bodypart(check_zone(selected_zone))

	var/datum/surgery/current_surgery

	for(var/s in living_target.surgeries)
		var/datum/surgery/surgery_datum = s
		if(S.location == selected_zone)
			current_surgery = surgery_datum

	if(!current_surgery)
		var/list/all_surgeries = GLOB.surgeries_list.Copy()
		var/list/available_surgeries = list()

		for(var/s in all_surgeries)
			var/datum/surgery/surgery_datum = s
			if(!surgery_datum.possible_locs.Find(selected_zone))
				continue
			if(affecting)
				if(!surgery_datum.requires_bodypart)
					continue
				if(surgery_datum.requires_bodypart_type && affecting.status != surgery_datum.requires_bodypart_type)
					continue
				if(surgery_datum.requires_real_bodypart && affecting.is_pseudopart)
					continue
			else if(carbon_target && surgery_datum.requires_bodypart) //mob with no limb in surgery zone when we need a limb
				continue
			if(surgery_datum.lying_required && (living_target.mobility_flags & MOBILITY_STAND))
				continue
			if(!surgery_datum.can_start(user, living_target))
				continue
			for(var/path in surgery_datum.target_mobtypes)
				if(istype(living_target, path))
					available_surgeries[surgery_datum.name] = surgery_datum
					break

		if(!length(available_surgeries))
			return

		var/input_procedure = input("Begin which procedure?", "Surgery", null, null) as null|anything in sortList(available_surgeries)
		if(input_procedure && user?.Adjacent(living_target) && (parent in user))
			var/datum/surgery/surgery_datum = available_surgeries[input_procedure]

			for(var/o in living_target.surgeries)
				var/datum/surgery/other = o
				if(other.location == selected_zone)
					return //during the input() another surgery was started at the same location.

			//we check that the surgery is still doable after the input() wait.
			if(carbon_target)
				affecting = carbon_target.get_bodypart(check_zone(selected_zone))
			if(affecting)
				if(!surgery_datum.requires_bodypart)
					return
				if(surgery_datum.requires_bodypart_type && affecting.status != surgery_datum.requires_bodypart_type)
					return
			else if(carbon_target && surgery_datum.requires_bodypart)
				return
			if(surgery_datum.lying_required && (living_target.mobility_flags & MOBILITY_STAND))
				return
			if(!surgery_datum.can_start(user, living_target))
				return

			if(surgery_datum.ignore_clothes || get_location_accessible(living_target, selected_zone))
				var/datum/surgery/procedure = new surgery_datum.type(living_target, selected_zone, affecting)
				user.visible_message("<span class='notice'>[user] drapes [parent] over [living_target]'s [parse_zone(selected_zone)] to prepare for surgery.</span>", \
					"<span class='notice'>You drape [parent] over [living_target]'s [parse_zone(selected_zone)] to prepare for \an [procedure.name].</span>")

				log_combat(user, living_target, "operated on", null, "(OPERATION TYPE: [procedure.name]) (TARGET AREA: [selected_zone])")
				after_select_cb?.Invoke()
			else
				to_chat(user, "<span class='warning'>You need to expose [living_target]'s [parse_zone(selected_zone)] first!</span>")

	else if(!current_surgery.step_in_progress)
		attempt_cancel_surgery(current_surgery, parent, living_target, user)
```

File: code/datums/components/surgery_initiator.dm [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+			to_chat(user, "<span class='warning'>You need to hold a [is_robotic ? "screwdriver" : "cautery"] in your inactive hand to stop [M]'s surgery!</span>")
+			return
+
+		if(S.operated_bodypart)
+			S.operated_bodypart.generic_bleedstacks -= 5
+
+		M.surgeries -= S
+		user.visible_message("<span class='notice'>[user] closes [M]'s [parse_zone(selected_zone)] with [close_tool] and removes [I].</span>", \
+			"<span class='notice'>You close [M]'s [parse_zone(selected_zone)] with [close_tool] and remove [I].</span>")
+		qdel(S)
```
Comment (Rohesie) [inline code/datums/components/surgery_initiator.dm:150]: Informative variables and early return.
```suggestion
/datum/component/surgery_initiator/proc/attempt_cancel_surgery(datum/surgery/surgery_datum, obj/item/surgery_item, mob/living/patient, mob/user)
	var/selected_zone = user.zone_selected

	if(surgery_datum.status == 1)
		patient.surgeries -= surgery_datum
		user.visible_message("<span class='notice'>[user] removes [surgery_item] from [patient]'s [parse_zone(selected_zone)].</span>", \
			"<span class='notice'>You remove [surgery_item] from [patient]'s [parse_zone(selected_zone)].</span>")
		qdel(surgery_datum)
		return

	if(!surgery_datum.can_cancel)
		return

	var/required_tool_type = TOOL_CAUTERY
	var/obj/item/close_tool = user.get_inactive_held_item()
	var/is_robotic = surgery_datum.requires_bodypart_type == BODYPART_ROBOTIC

	if(is_robotic)
		required_tool_type = TOOL_SCREWDRIVER

	if(iscyborg(user))
		close_tool = locate(/obj/item/cautery) in user.held_items
		if(!close_tool)
			to_chat(user, "<span class='warning'>You need to equip a cautery in an inactive slot to stop [patient]'s surgery!</span>")
			return
	else if(!close_tool || close_tool.tool_behaviour != required_tool_type)
		to_chat(user, "<span class='warning'>You need to hold a [is_robotic ? "screwdriver" : "cautery"] in your inactive hand to stop [patient]'s surgery!</span>")
		return

	if(surgery_datum.operated_bodypart)
		surgery_datum.operated_bodypart.generic_bleedstacks -= 5

	patient.surgeries -= surgery_datum
	user.visible_message("<span class='notice'>[user] closes [patient]'s [parse_zone(selected_zone)] with [close_tool] and removes [surgery_item].</span>", \
		"<span class='notice'>You close [patient]'s [parse_zone(selected_zone)] with [close_tool] and remove [surgery_item].</span>")
	qdel(surgery_datum)
```

**Comment by ExcessiveUseOfCobblestone** [comment]
i didnt commit the suggestion so may have missed something

File: code/datums/components/surgery_initiator.dm [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+				continue
+			for(var/path in surgeryloop_two.target_mobtypes)
+				if(istype(livingtarget, path))
+					available_surgeries[surgeryloop_two.name] = surgeryloop_two
+					break
+
+		if(!available_surgeries.len)
+			return
+
+		var/pick_your_surgery = input("Begin which procedure?", "Surgery", null, null) as null|anything in sortList(available_surgeries)
```
Comment (optimumtact) [inline code/datums/components/surgery_initiator.dm:82]: you cannot sleep in signal listeners, they are called with waitfor set to 0

Comment (optimumtact) [reply code/datums/components/surgery_initiator.dm:82]: this causes a race condition for the return value of the signal.

Comment (optimumtact) [reply code/datums/components/surgery_initiator.dm:82]: suggest you use getComponent and initiate a proc call directly on the component.

Comment (Rohesie) [reply code/datums/components/surgery_initiator.dm:82]: You can also invoke an async proc that sleeps for your timed action and guess a return value so things go on.
We should make signals async, really.

Comment (optimumtact) [reply code/datums/components/surgery_initiator.dm:82]: if you are both 
* sleeping in the proc
* and a signal

to me that's a sign you are trying to do too much over a signal boundary, and need to be using getcomponent instead.

Signals should not end up as a replacement for fucking proc calls

Comment (optimumtact) [reply code/datums/components/surgery_initiator.dm:82]: oh and also, that effectively makes every single send_signal a sleep, which means you need to recheck all your conditions after it.

Big bad imo

Comment (Rohesie) [reply code/datums/components/surgery_initiator.dm:82]: Well, potential one, yea.
Dunno, I see your point. There's pros and cons for either side.

Comment (ExcessiveUseOfCobblestone) [reply code/datums/components/surgery_initiator.dm:82]: so if I wanted to make potentially any obj/item a surgery initiator, I need to directly check for the component on obj/item attack/afterattack?

Comment (optimumtact) [reply code/datums/components/surgery_initiator.dm:82]: Or not use components for such a thing.

Comment (Rohesie) [reply code/datums/components/surgery_initiator.dm:82]: Do what I said before: once the `do_after()` begins invoke that as a separate async proc and return a value ending the attack chain

**Pull request merged**